File: scripts/serve-static.py
#!/usr/bin/env python3

from functools import partial
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from mimetypes import guess_type
from pathlib import Path
from sys import argv
from threading import Timer
from webbrowser import open as open_browser


HOST = "127.0.0.1"
FIRST_PORT = 4173
LAST_PORT = 4193
ROOT = Path(__file__).resolve().parent.parent / "网站成品"


class QuietHandler(SimpleHTTPRequestHandler):
    def log_message(self, format_string, *args):
        return

    def _requested_range(self, size):
        value = self.headers.get("Range", "")
        if not value.startswith("bytes=") or "," in value:
            return None
        start_text, separator, end_text = value[6:].partition("-")
        if not separator:
            return None
        try:
            if start_text:
                start = int(start_text)
                end = int(end_text) if end_text else size - 1
            elif end_text:
                suffix_length = int(end_text)
                start = max(size - suffix_length, 0)
                end = size - 1
            else:
                return None
        except ValueError:
            return None
        if start < 0 or end < start or start >= size:
            return None
        return start, min(end, size - 1)

    def _serve_range(self, head_only=False):
        if not self.headers.get("Range"):
            return False

        path = Path(self.translate_path(self.path))
        if not path.is_file():
            return False

        size = path.stat().st_size
        requested_range = self._requested_range(size)
        if requested_range is None:
            self.send_response(416)
            self.send_header("Content-Range", f"bytes */{size}")
            self.send_header("Accept-Ranges", "bytes")
            self.end_headers()
            return True

        start, end = requested_range
        self.send_response(206)
        self.send_header("Content-Type", guess_type(path.name)[0] or "application/octet-stream")
        self.send_header("Content-Length", str(end - start + 1))
        self.send_header("Content-Range", f"bytes {start}-{end}/{size}")
        self.send_header("Accept-Ranges", "bytes")
        self.send_header("Cache-Control", "public, max-age=3600")
        self.end_headers()

        if not head_only:
            remaining = end - start + 1
            with path.open("rb") as source:
                source.seek(start)
                while remaining:
                    chunk = source.read(min(64 * 1024, remaining))
                    if not chunk:
                        break
                    self.wfile.write(chunk)
                    remaining -= len(chunk)
        return True

    def do_GET(self):
        if not self._serve_range():
            super().do_GET()

    def do_HEAD(self):
        if not self._serve_range(head_only=True):
            super().do_HEAD()


def create_server():
    handler = partial(QuietHandler, directory=str(ROOT))
    for port in range(FIRST_PORT, LAST_PORT + 1):
        try:
            return ThreadingHTTPServer((HOST, port), handler), port
        except OSError:
            continue
    raise RuntimeError("端口 4173—4193 均被占用，请关闭其他本地网站后重试。")


def main():
    index_file = ROOT / "index.html"
    if not index_file.is_file():
        raise FileNotFoundError("未找到“网站成品/index.html”，请先运行 npm run build:static。")

    server, port = create_server()
    url = f"http://{HOST}:{port}"
    print(f"\n网站已启动：{url}")
    print("请保持此窗口开启；关闭窗口即可停止网站。\n")
    if "--no-open" not in argv:
        Timer(0.4, open_browser, args=(url,)).start()
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        pass
    finally:
        server.server_close()


if __name__ == "__main__":
    main()

File: scripts/serve-static.mjs
import { createReadStream } from "node:fs";
import { stat } from "node:fs/promises";
import { createServer } from "node:http";
import { extname, resolve, sep } from "node:path";
import { spawn } from "node:child_process";
import { fileURLToPath } from "node:url";

const host = "127.0.0.1";
const firstPort = 4173;
const lastPort = 4193;
const root = resolve(fileURLToPath(new URL("../网站成品/", import.meta.url)));
const openBrowser = !process.argv.includes("--no-open");

function normalizeSiteBasePath(value) {
  const trimmed = value?.trim();
  if (!trimmed || trimmed === "/") return "/";
  return `/${trimmed.replace(/^\/+|\/+$/g, "")}/`;
}

const siteBasePath = normalizeSiteBasePath(process.env.SITE_BASE_PATH);

const mimeTypes = new Map([
  [".css", "text/css; charset=utf-8"],
  [".html", "text/html; charset=utf-8"],
  [".js", "text/javascript; charset=utf-8"],
  [".json", "application/json; charset=utf-8"],
  [".svg", "image/svg+xml"],
  [".jpg", "image/jpeg"],
  [".jpeg", "image/jpeg"],
  [".png", "image/png"],
  [".webp", "image/webp"],
  [".mp4", "video/mp4"],
  [".woff2", "font/woff2"],
]);

function sendText(response, status, message) {
  response.writeHead(status, { "Content-Type": "text/plain; charset=utf-8" });
  response.end(message);
}

function browserCommand(url) {
  if (process.platform === "darwin") return ["open", [url]];
  if (process.platform === "win32") return ["cmd", ["/c", "start", "", url]];
  return ["xdg-open", [url]];
}

function openInBrowser(url) {
  const [command, args] = browserCommand(url);
  const child = spawn(command, args, {
    detached: true,
    stdio: "ignore",
  });
  child.unref();
}

async function resolveRequestPath(pathname) {
  let sitePath = pathname;
  if (siteBasePath !== "/") {
    const baseWithoutTrailingSlash = siteBasePath.slice(0, -1);
    if (pathname === baseWithoutTrailingSlash) sitePath = "/";
    else if (pathname.startsWith(siteBasePath)) {
      sitePath = pathname.slice(baseWithoutTrailingSlash.length) || "/";
    } else return null;
  }

  let decoded;
  try {
    decoded = decodeURIComponent(sitePath);
  } catch {
    return null;
  }

  const relativePath = decoded.replace(/^\/+/, "") || "index.html";
  let filePath = resolve(root, relativePath);
  if (filePath !== root && !filePath.startsWith(`${root}${sep}`)) return null;

  try {
    const fileStat = await stat(filePath);
    if (fileStat.isDirectory()) filePath = resolve(filePath, "index.html");
    return { filePath, fileStat: await stat(filePath) };
  } catch {
    if (!extname(relativePath)) {
      const fallback = resolve(root, "index.html");
      return { filePath: fallback, fileStat: await stat(fallback) };
    }
    return null;
  }
}

function parseRange(rangeHeader, size) {
  const match = /^bytes=(\d*)-(\d*)$/.exec(rangeHeader ?? "");
  if (!match) return null;

  let start = match[1] ? Number(match[1]) : 0;
  let end = match[2] ? Number(match[2]) : size - 1;

  if (!match[1] && match[2]) {
    const suffixLength = Number(match[2]);
    start = Math.max(size - suffixLength, 0);
    end = size - 1;
  }

  if (!Number.isInteger(start) || !Number.isInteger(end) || start < 0 || end < start || start >= size) {
    return null;
  }

  return { start, end: Math.min(end, size - 1) };
}

const server = createServer(async (request, response) => {
  if (request.method !== "GET" && request.method !== "HEAD") {
    response.setHeader("Allow", "GET, HEAD");
    sendText(response, 405, "仅支持读取网站资源。");
    return;
  }

  const url = new URL(request.url ?? "/", `http://${host}`);
  const resolved = await resolveRequestPath(url.pathname);
  if (!resolved) {
    sendText(response, 404, "未找到该资源。");
    return;
  }

  const { filePath, fileStat } = resolved;
  const contentType = mimeTypes.get(extname(filePath).toLowerCase()) ?? "application/octet-stream";
  const headers = {
    "Accept-Ranges": "bytes",
    "Cache-Control": filePath.endsWith("index.html") ? "no-cache" : "public, max-age=3600",
    "Content-Type": contentType,
  };
  const range = parseRange(request.headers.range, fileStat.size);

  if (request.headers.range && !range) {
    response.writeHead(416, {
      ...headers,
      "Content-Range": `bytes */${fileStat.size}`,
    });
    response.end();
    return;
  }

  if (range) {
    response.writeHead(206, {
      ...headers,
      "Content-Length": range.end - range.start + 1,
      "Content-Range": `bytes ${range.start}-${range.end}/${fileStat.size}`,
    });
    if (request.method === "HEAD") response.end();
    else createReadStream(filePath, range).pipe(response);
    return;
  }

  response.writeHead(200, {
    ...headers,
    "Content-Length": fileStat.size,
  });
  if (request.method === "HEAD") response.end();
  else createReadStream(filePath).pipe(response);
});

server.on("clientError", (_error, socket) => {
  socket.end("HTTP/1.1 400 Bad Request\r\n\r\n");
});

function listen(port) {
  const onError = (error) => {
    server.off("listening", onListening);
    if (error.code === "EADDRINUSE" && port < lastPort) {
      listen(port + 1);
      return;
    }
    console.error(`网站启动失败：${error.message}`);
    process.exitCode = 1;
  };

  const onListening = () => {
    server.off("error", onError);
    const url = `http://${host}:${port}${siteBasePath === "/" ? "" : siteBasePath}`;
    console.log(`\n网站已启动：${url}`);
    console.log("请保持此窗口开启；关闭窗口即可停止网站。\n");
    if (openBrowser) openInBrowser(url);
  };

  server.once("error", onError);
  server.once("listening", onListening);
  server.listen(port, host);
}

try {
  await stat(resolve(root, "index.html"));
  listen(firstPort);
} catch {
  console.error("未找到“网站成品/index.html”，请先运行 npm run build:static。");
  process.exitCode = 1;
}
